src/Microsoft.AspNetCore.SignalR.Core/Internal/HubMethodDescriptor.cs

```diff
@@ -0,0 +1,149 @@
+
+        public bool IsObservable { get; }
+
+        public bool IsChannel { get; }
+
+        public bool IsStreamable => IsObservable || IsChannel;
+
+        public Type StreamReturnType { get; }
+
+        public IList<IAuthorizeData> Policies { get; }
+
+        private static bool IsChannelType(Type type, out Type payloadType)
+        {
+            var channelType = type.AllBaseTypes().FirstOrDefault(t => t.IsGenericType && t.GetGenericTypeDefinition() == typeof(ChannelReader<>));
+            if (channelType == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = channelType.GetGenericArguments()[0];
+            return true;
+        }
+
+        private static bool IsObservableType(Type type, out Type payloadType)
+        {
+            var observableInterface = IsIObservable(type) ? type : type.GetInterfaces().FirstOrDefault(IsIObservable);
+            if (observableInterface == null)
+            {
+                payloadType = null;
+                return false;
+            }
+
+            payloadType = observableInterface.GetGenericArguments()[0];
+            return true;
+
+            bool IsIObservable(Type iface)
+            {
+                return iface.IsGenericType && iface.GetGenericTypeDefinition() == typeof(IObservable<>);
+            }
+        }
+
+        public IAsyncEnumerator<object> FromObservable(object observable, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CompileConvertToEnumerator(FromObservableMethod);
+            }
+
+            return _convertToEnumerator.Invoke(observable, cancellationToken);
+        }
+
+        public IAsyncEnumerator<object> FromChannel(object channel, CancellationToken cancellationToken)
+        {
+            if (_convertToEnumerator == null)
+            {
+                _convertToEnumerator = CompileConvertToEnumerator(GetAsyncEnumeratorMethod);
+            }
+
+            return _convertToEnumerator.Invoke(channel, cancellationToken);
+        }
+
+        private Func<object, CancellationToken, IAsyncEnumerator<object>> CompileConvertToEnumerator(MethodInfo methodInfo)
+        {
+            // This will call one of two methods to wrap the passed in streamable value
+            // and cancellation token to an IAsyncEnumerator<object>
+            //
+            // IObservable<T>:
+            // AsyncEnumeratorAdapters.FromObservable<T>(observable, cancellationToken);
+            //
+            // ChannelReader<T>
+            // AsyncEnumeratorAdapters.GetAsyncEnumerator<T>(channelReader, cancellationToken);
+
+            var genericMethodInfo = methodInfo.MakeGenericMethod(StreamReturnType);
+
+            var methodParameters = genericMethodInfo.GetParameters();
+
+            var targetParameter = Expression.Parameter(typeof(object), "arg1");
```

Comment:
names for the arguments? I don't know if they show up in callstacks or something but if `Expression.Parameter` takes a name, perhaps it should be a more useful name than `arg1`. That's assuming compiling to `Func` doesn't mandate the argument names. Even if they get wiped out by compiling to `Func`, it doesn't hurt, if nothing else it serves as commentary in this code

Reply:
`arg1` and `arg2` is coming from the parameter names of `Func<T1, T2, TReturn1>`. Need to use those names to get a reference to them and pass them to our own method. I'll improve docs to be clearer